File: shuffle_mcqs.py
import random
import re

def shuffle_mcqs(input_file, output_file):
    """
    Shuffle MCQ options while maintaining correct answer relationships.
    """
    with open(input_file, 'r', encoding='utf-8') as f:
        content = f.read()
    
    # Split content into questions
    questions = content.strip().split('\n\n')
    shuffled_questions = []
    
    for question_block in questions:
        if not question_block.strip():
            continue
            
        lines = question_block.strip().split('\n')
        if len(lines) < 6:  # Need at least question + 4 options + answer
            shuffled_questions.append(question_block)
            continue
        
        # Extract question text
        question_text = lines[0]
        
        # Extract options
        options = []
        answer_line = None
        
        for line in lines[1:]:
            if line.startswith('Answer:'):
                answer_line = line
                break
            elif re.match(r'^[A-D]\)', line):
                options.append(line)
        
        if len(options) != 4 or not answer_line:
            shuffled_questions.append(question_block)
            continue
        
        # Extract current answer
        current_answer = answer_line.split(':')[1].strip()
        
        # Create option pairs (letter, text)
        option_pairs = []
        for option in options:
            letter = option[0]
            text = option[3:]  # Remove "A) " part
            option_pairs.append((letter, text))
        
        # Shuffle options
        random.shuffle(option_pairs)
        
        # Find the correct answer text
        correct_text = None
        for letter, text in option_pairs:
            if letter == current_answer:
                correct_text = text
                break
        
        # Find new position of correct answer
        new_answer = None
        for i, (letter, text) in enumerate(option_pairs):
            if text == correct_text:
                new_answer = chr(65 + i)  # Convert 0,1,2,3 to A,B,C,D
                break
        
        # Reconstruct question
        new_question_lines = [question_text]
        for i, (letter, text) in enumerate(option_pairs):
            new_letter = chr(65 + i)
            new_question_lines.append(f"{new_letter}) {text}")
        
        new_question_lines.append(f"Answer: {new_answer}")
        
        shuffled_questions.append('\n'.join(new_question_lines))
    
    # Write shuffled questions to output file
    with open(output_file, 'w', encoding='utf-8') as f:
        f.write('\n\n'.join(shuffled_questions))
    
    print(f"Shuffled MCQs saved to {output_file}")

if __name__ == "__main__":
    # Set random seed for reproducibility (optional)
    random.seed(42)
    
    # Shuffle the MCQs
    shuffle_mcqs('all_mcqs.txt', 'all_mcqs_shuffled.txt') 

File: shuffle_mcqs.js
const fs = require('fs');

function shuffleArray(array) {
    for (let i = array.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [array[i], array[j]] = [array[j], array[i]];
    }
    return array;
}

function shuffleMCQs(inputFile, outputFile) {
    try {
        // Read the input file
        const content = fs.readFileSync(inputFile, 'utf8');
        
        // Split content into questions
        const questions = content.trim().split('\n\n');
        const shuffledQuestions = [];
        
        for (const questionBlock of questions) {
            if (!questionBlock.trim()) continue;
            
            const lines = questionBlock.trim().split('\n');
            if (lines.length < 6) { // Need at least question + 4 options + answer
                shuffledQuestions.push(questionBlock);
                continue;
            }
            
            // Extract question text
            const questionText = lines[0];
            
            // Extract options and answer
            const options = [];
            let answerLine = null;
            
            for (let i = 1; i < lines.length; i++) {
                const line = lines[i];
                if (line.startsWith('Answer:')) {
                    answerLine = line;
                    break;
                } else if (/^[A-D]\)/.test(line)) {
                    options.push(line);
                }
            }
            
            if (options.length !== 4 || !answerLine) {
                shuffledQuestions.push(questionBlock);
                continue;
            }
            
            // Extract current answer
            const currentAnswer = answerLine.split(':')[1].trim();
            
            // Create option pairs (letter, text)
            const optionPairs = options.map(option => {
                const letter = option[0];
                const text = option.substring(3); // Remove "A) " part
                return { letter, text };
            });
            
            // Shuffle options
            shuffleArray(optionPairs);
            
            // Find the correct answer text
            const correctOption = optionPairs.find(pair => pair.letter === currentAnswer);
            const correctText = correctOption ? correctOption.text : null;
            
            // Find new position of correct answer
            const newAnswerIndex = optionPairs.findIndex(pair => pair.text === correctText);
            const newAnswer = newAnswerIndex !== -1 ? String.fromCharCode(65 + newAnswerIndex) : currentAnswer;
            
            // Reconstruct question
            const newQuestionLines = [questionText];
            optionPairs.forEach((pair, index) => {
                const newLetter = String.fromCharCode(65 + index);
                newQuestionLines.push(`${newLetter}) ${pair.text}`);
            });
            newQuestionLines.push(`Answer: ${newAnswer}`);
            
            shuffledQuestions.push(newQuestionLines.join('\n'));
        }
        
        // Write shuffled questions to output file
        fs.writeFileSync(outputFile, shuffledQuestions.join('\n\n'), 'utf8');
        console.log(`Shuffled MCQs saved to ${outputFile}`);
        
    } catch (error) {
        console.error('Error processing file:', error.message);
    }
}

// Set random seed for reproducibility (optional)
Math.seedrandom = function(seed) {
    let x = seed;
    return function() {
        x = (x * 9301 + 49297) % 233280;
        return x / 233280;
    };
};

// Use a fixed seed for reproducible results
const random = Math.seedrandom(42);
Math.random = random;

// Shuffle the MCQs
shuffleMCQs('all_mcqs.txt', 'all_mcqs_shuffled.txt'); 